mlflow/utils/stdout_logging.py

```diff
@@ -0,0 +1,95 @@
+import sys
+import threading
+import time
+from contextlib import contextmanager
+from io import StringIO
+
+import mlflow
+
+
+class TeeStringIO:
+    """A file-like object that writes to both original stdout and a StringIO buffer."""
+
+    def __init__(self, original_stdout, string_buffer):
+        self.original_stdout = original_stdout
+        self.string_buffer = string_buffer
+
+    def write(self, data):
+        # Write to both original stdout and our buffer
+        self.original_stdout.write(data)
+        self.string_buffer.write(data)
+        return len(data)
+
+    def flush(self):
+        self.original_stdout.flush()
+        self.string_buffer.flush()
+
+    def __getattr__(self, name):
+        # Delegate other attributes to original stdout
+        return getattr(self.original_stdout, name)
+
+
+@contextmanager
+def log_stdout_stream(run_id, interval_seconds=5):
+    """
+    A context manager to stream stdout to an MLflow artifact.
+
+    This context manager redirects `sys.stdout` to an in-memory buffer.
+    A background thread periodically flushes this buffer and logs its
+    contents to an MLflow artifact file named 'stdout.log'.
+
+    Args:
+        run_id (str): The run ID to log stdout to.
+        interval_seconds (int): The interval in seconds at which to log
+                                the stdout buffer to MLflow.
+
+    Example:
+        import time
+        import mlflow
+
+        with mlflow.start_run() as run:
+            with log_stdout_stream(run.info.run_id):
+                print("This is the start of my script.")
+                time.sleep(6)
+                print("This message will appear in the first log upload.")
+                time.sleep(6)
+                print("And this will be in the second.")
+            # The context manager will automatically handle final log upload
+            # and cleanup.
+        print("Stdout is now back to normal.")
+    """
+    original_stdout = sys.stdout
+    stdout_buffer = StringIO()
+    tee_stdout = TeeStringIO(original_stdout, stdout_buffer)
+    sys.stdout = tee_stdout
```

Comment:
this is nice but it won't log the stdout of underlying e.g. C or C++ libraries. only python. To log that, you need to reopen filedescriptor 1 and set it to a pipe (and then have a pipe callback which writes to this `stdout_buffer` or similar)